zerver/tornado/event_queue.py

```diff
@@ -130,6 +130,15 @@ def from_dict(cls, d: MutableMapping[str, Any]) -> 'ClientDescriptor':
         if 'client_gravatar' not in d:
             # Temporary migration for the addition of the client_gravatar field
             d['client_gravatar'] = False
+        if 'newest_pruned_id' not in d['event_queue']:
+            # Temporary migration for the addition of the newest_pruned_id field
+            if len(d['event_queue']['queue']) > 0 or len(d['event_queue']['virtual_events']) > 0:
+                d['event_queue']['newest_pruned_id'] = min(
+                    [event['id'] for event in d['event_queue']['queue']] +
+                    [event['id'] for event in d['event_queue']['virtual_events'].values()],
+                ) - 1
+            else:
+                d['event_queue']['newest_pruned_id'] = d['next_event_id'] - 1
```

Comment:
You meant `d['next_event_id']` ↦ `d['event_queue']['next_event_id']`. (Why is this in `ClientDescriptor.from_dict` instead of `EventQueue.from_dict`?)

However, I think this logic is still wrong in the case where the earliest event after the last fetch was a virtual event that was then replaced with a later virtual event of the same type?